### PR TITLE
[NeMo-UX] Added Megatron loss function and LR scheduler

### PR DESCRIPTION
**Pull request by ahmadki**
# What does this PR do ?

The PR adds Megatron loss function and LR scheduler for GPT pretrain.

Add a one line overview of what this PR aims to accomplish.

**Collection**: [Note which collection this PR will affect]

LLM

# Changelog 
- Parametrized GPT model loss function
- Added `MegatronAnnealing` learning rate and weight decay scheduler
- Added `MegatronMaskedTokenLossReduction` loss function

# Usage
* You can potentially add a usage example below

```python
# Add a code snippet demonstrating how to use this 
```

# GitHub Actions CI

The Jenkins CI system has been replaced by GitHub Actions self-hosted runners.

The GitHub Actions CI will run automatically when the "Run CICD" label is added to the PR.
To re-run CI remove and add the label again.
To run CI on an untrusted fork, a NeMo user with write access must first click "Approve and run".

# Before your PR is "Ready for review"
**Pre checks**:
- [ ] Make sure you read and followed [Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md)
- [ ] Did you write any new necessary tests?
- [ ] Did you add or update any necessary documentation?
- [ ] Does the PR affect components that are optional to install? (Ex: Numba, Pynini, Apex etc)
  - [ ] Reviewer: Does the PR have correct import guards for all optional libraries?
  
**PR Type**:
- [ x] New Feature
- [ x] Bugfix
- [ ] Documentation

If you haven't finished some of the above items you can still open "Draft" PR.

## Who can review?

Anyone in the community is free to review the PR once the checks have passed. 
[Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md) contains specific people who can review PRs to various areas.

# Additional Information
* Related to # (issue)

### CODE REVIEW

File: nemo/core/optim/lr_scheduler.py [hunk 1]
```diff
@@ -665,6 +665,171 @@ def _get_lr(self, step):
...
+        lr_decay_style,
+        start_wd,
+        end_wd,
+        wd_incr_steps,
+        wd_incr_style,
+        use_checkpoint_opt_param_scheduler=True,
+        override_opt_param_scheduler=False,
+        last_epoch=-1,
+    ):
+        """
```
Comment (marcromeyn) [inline nemo/core/optim/lr_scheduler.py:689]: I think the rest of the classes put the doc-string right below `class MegatronAnnealing`

File: nemo/lightning/pytorch/optim/lr_scheduler.py [hunk 2]
```diff
@@ -461,3 +462,64 @@ def scheduler(self, model, optimizer):
             # Metric to to monitor for schedulers like `ReduceLROnPlateau`
             "monitor": self.monitor,
         }
+
+
+class MegatronAnnealingScheduler(LRSchedulerModule):
```
Comment (marcromeyn) [inline nemo/lightning/pytorch/optim/lr_scheduler.py:467]: Could use a doc-string (ideally with an example inside it)

**Review by ashors1 — APPROVED** [review]
